File: test/unit/limits.test.js
import { describe, expect, it } from "vitest";
import {
  FLAG_ABS,
  FLAG_TC,
  FLAG_VSC,
  LIMIT_KINDS,
  MERGE_GAP_POINTS,
  activeLimitLabels,
  booleanRuns,
  hasLimitData,
  limitAt,
  limitMarkers,
  limitRuns,
  limitSummary,
  sessionLimits,
} from "../../public/js/limits.js";

// 20 grid points: ABS pulses across a braking zone (k 2–5, with a one-point
// gap), TC once on an exit (k 9–10), VSC never; wheelspin on that same exit
// and a lockup at k 3.
const flags = [0, 0, FLAG_ABS, 0, FLAG_ABS, FLAG_ABS, 0, 0, 0, FLAG_TC, FLAG_TC, 0, 0, 0, 0, 0, 0, 0, 0, 0];
const slip = [0, 0, 0, -3, -1, 0, 0, 0, 0.5, 3, 4.5, 1, 0, 0, 0, 0, 0, 0, 0, 0];
const lap = { n: 1, timeMs: 90_000, speed: Array(20).fill(100), flags, wheelSlip: slip };

describe("limitAt / hasLimitData", () => {
  it("reads the flag bits and the slip thresholds, null for a channel the lap lacks", () => {
    expect(limitAt(lap, "abs", 2)).toBe(true);
    expect(limitAt(lap, "abs", 3)).toBe(false);
    expect(limitAt(lap, "tc", 9)).toBe(true);
    expect(limitAt(lap, "vsc", 9)).toBe(false);
    expect(limitAt(lap, "wheelspin", 10)).toBe(true);
    expect(limitAt(lap, "wheelspin", 8)).toBe(false); // 0.5 % is noise
    expect(limitAt(lap, "lockup", 3)).toBe(true);
    expect(limitAt({ flags }, "wheelspin", 3)).toBeNull();
    expect(limitAt({ wheelSlip: slip }, "abs", 3)).toBeNull();
    expect(limitAt(lap, "abs", 99)).toBeNull();
    expect(hasLimitData(lap)).toBe(true);
    expect(hasLimitData({ speed: [1, 2] })).toBe(false);
  });
  it("VSC is bit 2", () => {
    expect(limitAt({ flags: [FLAG_VSC | FLAG_ABS] }, "vsc", 0)).toBe(true);
    expect(limitAt({ flags: [FLAG_VSC | FLAG_ABS] }, "tc", 0)).toBe(false);
  });
});

describe("booleanRuns", () => {
  it("merges runs across short gaps and keeps longer ones apart", () => {
    const s = [false, true, true, false, true, false, false, false, true, true];
    expect(booleanRuns(s, 2)).toEqual([
      { k0: 1, k1: 4 }, // one clear point between: merged
      { k0: 8, k1: 9 }, // three clear points: separate
    ]);
    expect(booleanRuns(s, 0)).toEqual([{ k0: 1, k1: 2 }, { k0: 4, k1: 4 }, { k0: 8, k1: 9 }]);
    expect(booleanRuns([], 2)).toEqual([]);
    expect(MERGE_GAP_POINTS).toBe(2);
  });
});

describe("limitRuns", () => {
  it("lists every kind's runs in kind order, the ABS pulse train as one run", () => {
    expect(limitRuns(lap)).toEqual([
      { kind: "abs", k0: 2, k1: 5 },
      { kind: "lockup", k0: 3, k1: 3 },
      { kind: "tc", k0: 9, k1: 10 },
      { kind: "wheelspin", k0: 9, k1: 10 },
    ]);
  });
  it("skips the kinds whose channel is missing", () => {
    expect(limitRuns({ flags }).map((r) => r.kind)).toEqual(["abs", "tc"]);
    expect(limitRuns({ speed: [1, 2] })).toEqual([]);
  });
});

describe("activeLimitLabels", () => {
  it("names what is active at a grid point", () => {
    expect(activeLimitLabels(lap, 3)).toEqual(["Lockup"]);
    expect(activeLimitLabels(lap, 10)).toEqual(["Traction control", "Wheelspin"]);
    expect(activeLimitLabels(lap, 0)).toEqual([]);
  });
});

describe("sessionLimits / limitSummary", () => {
  const lap2 = {
    n: 2,
    timeMs: 91_000,
    speed: Array(20).fill(100),
    // ABS in the same zone, plus a second zone at k 15–16; no slip channel
    flags: flags.map((f, k) => (k === 15 || k === 16 ? FLAG_ABS : f & FLAG_ABS)),
  };

  it("counts distinct places across laps and the laps involved", () => {
    const sl = sessionLimits({ v: 1, dStepM: 20, laps: [lap, lap2, { n: 3, speed: [1, 2] }] });
    expect(sl.hasFlags).toBe(true);
    expect(sl.hasSlip).toBe(true);
    expect(sl.kinds).toEqual([
      { kind: "abs", places: 2, laps: 2 },
      { kind: "lockup", places: 1, laps: 1 },
      { kind: "tc", places: 1, laps: 1 },
      { kind: "wheelspin", places: 1, laps: 1 },
      { kind: "vsc", places: 0, laps: 0 },
    ]);
    expect(limitSummary({ v: 1, dStepM: 20, laps: [lap, lap2] })).toBe(
      "ABS in 2 braking zones, lockup in 1 braking zone, traction control in 1 acceleration zone, wheelspin in 1 acceleration zone"
    );
  });

  it("says no interventions when the systems never fired, and null without the channels", () => {
    const quiet = { v: 1, dStepM: 20, laps: [{ n: 1, flags: Array(20).fill(0) }] };
    expect(limitSummary(quiet)).toBe("no interventions");
    expect(sessionLimits(quiet).kinds.map((k) => k.kind)).toEqual(["abs", "tc", "vsc"]); // no slip channel: those kinds absent
    expect(limitSummary({ v: 1, dStepM: 20, laps: [{ n: 1, speed: [1, 2] }] })).toBeNull();
    expect(limitSummary(null)).toBeNull();
  });
});

describe("limitMarkers", () => {
  // A 380 m straight line trace sampled every 10 m, so distance fractions
  // map to indexes directly.
  const trace = Array.from({ length: 39 }, (_, i) => [i * 10, 0, 50]);

  it("places each run's mid-point on the trace by driven-distance fraction", () => {
    const m = limitMarkers(lap, 20, trace);
    expect(m.map((x) => x.kind)).toEqual(["abs", "lockup", "tc", "wheelspin"]);
    // ABS mid-point k=3.5 of 19 → 70 m of 380 → index 7 on the 10 m trace.
    expect(m[0]).toEqual({ kind: "abs", k0: 2, k1: 5, idx: 7 });
    expect(m[2].idx).toBe(19); // k=9.5 → 190 m
  });

  it("scales to the trace's own length when it differs from the grid length", () => {
    const shortTrace = Array.from({ length: 20 }, (_, i) => [i * 10, 0, 50]); // 190 m
    expect(limitMarkers(lap, 20, shortTrace)[0].idx).toBe(4); // 70/380 · 190 = 35 m → first point ≥ 35 m
  });

  it("is empty without a trace or without runs", () => {
    expect(limitMarkers(lap, 20, null)).toEqual([]);
    expect(limitMarkers({ speed: Array(20).fill(100), flags: Array(20).fill(0) }, 20, trace)).toEqual([]);
  });
});

describe("LIMIT_KINDS", () => {
  it("colours by side and never leaves two kinds of one side colour-alone", () => {
    for (const side of ["brake", "power"]) {
      const kinds = LIMIT_KINDS.filter((k) => k.side === side);
      expect(kinds).toHaveLength(2);
      expect(kinds[0].filled).not.toBe(kinds[1].filled);
    }
  });
});

File: apps/android/core/src/test/kotlin/app/trackevolution/core/LimitsTest.kt
package app.trackevolution.core

import app.trackevolution.core.model.LapChannels
import app.trackevolution.core.model.SessionChannels
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * The port of `test/unit/limits.test.js`, case for case, plus the
 * cross-language pin against `contracts/logic/limits.json`.
 */
class LimitsTest {

    private val abs = Limits.FLAG_ABS.toDouble()
    private val tc = Limits.FLAG_TC.toDouble()
    private val vsc = Limits.FLAG_VSC.toDouble()

    /**
     * 20 grid points: ABS pulses across a braking zone (k 2–5, with a one-point
     * gap), TC once on an exit (k 9–10), VSC never; wheelspin on that same exit
     * and a lockup at k 3.
     */
    private val flags = listOf(
        0.0, 0.0, abs, 0.0, abs, abs, 0.0, 0.0, 0.0, tc,
        tc, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
    )
    private val slip = listOf(
        0.0, 0.0, 0.0, -3.0, -1.0, 0.0, 0.0, 0.0, 0.5, 3.0,
        4.5, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
    )
    private val lap = LapChannels(
        n = 1,
        timeMs = 90_000,
        speed = List(20) { 100.0 },
        wheelSlip = slip,
        flags = flags,
    )

    // ---- limitAt / hasLimitData -------------------------------------------

    @Test
    fun `reads the flag bits and the slip thresholds, null for a channel the lap lacks`() {
        assertEquals(true, Limits.limitAt(lap, "abs", 2))
        assertEquals(false, Limits.limitAt(lap, "abs", 3))
        assertEquals(true, Limits.limitAt(lap, "tc", 9))
        assertEquals(false, Limits.limitAt(lap, "vsc", 9))
        assertEquals(true, Limits.limitAt(lap, "wheelspin", 10))
        assertEquals(false, Limits.limitAt(lap, "wheelspin", 8)) // 0.5 % is noise
        assertEquals(true, Limits.limitAt(lap, "lockup", 3))
        assertNull(Limits.limitAt(LapChannels(n = 1, timeMs = 0, flags = flags), "wheelspin", 3))
        assertNull(Limits.limitAt(LapChannels(n = 1, timeMs = 0, wheelSlip = slip), "abs", 3))
        assertNull(Limits.limitAt(lap, "abs", 99))
        assertTrue(Limits.hasLimitData(lap))
        assertFalse(Limits.hasLimitData(LapChannels(n = 1, timeMs = 0, speed = listOf(1.0, 2.0))))
    }

    @Test
    fun `VSC is bit 2`() {
        val both = LapChannels(n = 1, timeMs = 0, flags = listOf(vsc + abs))
        assertEquals(true, Limits.limitAt(both, "vsc", 0))
        assertEquals(false, Limits.limitAt(both, "tc", 0))
    }

    // ---- booleanRuns -------------------------------------------------------

    @Test
    fun `merges runs across short gaps and keeps longer ones apart`() {
        val s = listOf(false, true, true, false, true, false, false, false, true, true)
        assertEquals(
            listOf(
                Limits.Run(1, 4), // one clear point between: merged
                Limits.Run(8, 9), // three clear points: separate
            ),
            Limits.booleanRuns(s, 2),
        )
        assertEquals(
            listOf(Limits.Run(1, 2), Limits.Run(4, 4), Limits.Run(8, 9)),
            Limits.booleanRuns(s, 0),
        )
        assertEquals(emptyList<Limits.Run>(), Limits.booleanRuns(emptyList(), 2))
        assertEquals(2, Limits.MERGE_GAP_POINTS)
    }

    // ---- limitRuns ---------------------------------------------------------

    @Test
    fun `lists every kind's runs in kind order, the ABS pulse train as one run`() {
        assertEquals(
            listOf(
                Limits.LimitRun("abs", 2, 5),
                Limits.LimitRun("lockup", 3, 3),
                Limits.LimitRun("tc", 9, 10),
                Limits.LimitRun("wheelspin", 9, 10),
            ),
            Limits.limitRuns(lap),
        )
    }

    @Test
    fun `skips the kinds whose channel is missing`() {
        assertEquals(
            listOf("abs", "tc"),
            Limits.limitRuns(LapChannels(n = 1, timeMs = 0, flags = flags)).map { it.kind },
        )
        assertEquals(
            emptyList<Limits.LimitRun>(),
            Limits.limitRuns(LapChannels(n = 1, timeMs = 0, speed = listOf(1.0, 2.0))),
        )
    }

    // ---- activeLimitLabels -------------------------------------------------

    @Test
    fun `names what is active at a grid point`() {
        assertEquals(listOf("Lockup"), Limits.activeLimitLabels(lap, 3))
        assertEquals(listOf("Traction control", "Wheelspin"), Limits.activeLimitLabels(lap, 10))
        assertEquals(emptyList<String>(), Limits.activeLimitLabels(lap, 0))
    }

    // ---- sessionLimits / limitSummary --------------------------------------

    /** ABS in the same zone, plus a second zone at k 15–16; no slip channel. */
    private val lap2 = LapChannels(
        n = 2,
        timeMs = 91_000,
        speed = List(20) { 100.0 },
        flags = flags.mapIndexed { k, f ->
            if (k == 15 || k == 16) abs else (f.toInt() and Limits.FLAG_ABS).toDouble()
        },
    )

    @Test
    fun `counts distinct places across laps and the laps involved`() {
        val sl = Limits.sessionLimits(
            SessionChannels(
                v = 1,
                dStepM = 20.0,
                laps = listOf(lap, lap2, LapChannels(n = 3, timeMs = 0, speed = listOf(1.0, 2.0))),
            ),
        )!!
        assertTrue(sl.hasFlags)
        assertTrue(sl.hasSlip)
        assertEquals(
            listOf(
                Limits.KindTally("abs", 2, 2),
                Limits.KindTally("lockup", 1, 1),
                Limits.KindTally("tc", 1, 1),
                Limits.KindTally("wheelspin", 1, 1),
                Limits.KindTally("vsc", 0, 0),
            ),
            sl.kinds,
        )
        assertEquals(
            "ABS in 2 places, lockup in 1 place, traction control in 1 place, wheelspin in 1 place",
            Limits.limitSummary(SessionChannels(v = 1, dStepM = 20.0, laps = listOf(lap, lap2))),
        )
    }

    @Test
    fun `says no interventions when the systems never fired, and null without the channels`() {
        val quiet = SessionChannels(
            v = 1,
            dStepM = 20.0,
            laps = listOf(LapChannels(n = 1, timeMs = 0, flags = List(20) { 0.0 })),
        )
        assertEquals("no interventions", Limits.limitSummary(quiet))
        // No slip channel: those kinds are absent rather than reported as zero.
        assertEquals(listOf("abs", "tc", "vsc"), Limits.sessionLimits(quiet)!!.kinds.map { it.kind })
        assertNull(
            Limits.limitSummary(
                SessionChannels(
                    v = 1,
                    dStepM = 20.0,
                    laps = listOf(LapChannels(n = 1, timeMs = 0, speed = listOf(1.0, 2.0))),
                ),
            ),
        )
        assertNull(Limits.limitSummary(SessionChannels(v = 1, dStepM = 20.0, laps = emptyList())))
    }

    // ---- limitMarkers ------------------------------------------------------

    /**
     * A 380 m straight-line trace sampled every 10 m, so distance fractions map
     * to indexes directly.
     */
    private val trace = List(39) { TraceSample(x = it * 10.0, y = 0.0, v = 50.0) }

    @Test
    fun `places each run's mid-point on the trace by driven-distance fraction`() {
        val m = Limits.limitMarkers(lap, 20.0, trace)
        assertEquals(listOf("abs", "lockup", "tc", "wheelspin"), m.map { it.kind })
        // ABS mid-point k=3.5 of 19 → 70 m of 380 → index 7 on the 10 m trace.
        assertEquals(Limits.Marker("abs", 2, 5, 7), m[0])
        assertEquals(19, m[2].idx) // k=9.5 → 190 m
    }

    @Test
    fun `scales to the trace's own length when it differs from the grid length`() {
        val shortTrace = List(20) { TraceSample(x = it * 10.0, y = 0.0, v = 50.0) } // 190 m
        // 70/380 · 190 = 35 m → the first point at or past 35 m.
        assertEquals(4, Limits.limitMarkers(lap, 20.0, shortTrace)[0].idx)
    }

    @Test
    fun `markers are empty without a trace or without runs`() {
        assertEquals(emptyList<Limits.Marker>(), Limits.limitMarkers(lap, 20.0, null))
        assertEquals(
            emptyList<Limits.Marker>(),
            Limits.limitMarkers(
                LapChannels(n = 1, timeMs = 0, speed = List(20) { 100.0 }, flags = List(20) { 0.0 }),
                20.0,
                trace,
            ),
        )
    }

    // ---- LIMIT_KINDS -------------------------------------------------------

    @Test
    fun `colours by side and never leaves two kinds of one side colour-alone`() {
        for (side in listOf(Limits.Side.BRAKE, Limits.Side.POWER)) {
            val kinds = Limits.LIMIT_KINDS.filter { it.side == side }
            assertEquals(2, kinds.size)
            assertNotEquals(kinds[0].filled, kinds[1].filled)
        }
    }

    // ---- cross-language pin ------------------------------------------------

    /**
     * The JS implementation's own output for a shared input has to come back out
     * of this port integer for integer and string for string.
     */
    @Test
    fun `matches the JavaScript implementation on a shared fixture`() {
        val json = Json { ignoreUnknownKeys = false }
        val fixture = Json.parseToJsonElement(
            RepoRoot.path("contracts/logic/limits.json").readText(),
        ).jsonObject
        val input = fixture["input"]!!.jsonObject
        val expected = fixture["expected"]!!.jsonObject
        val channels = json.decodeFromJsonElement<SessionChannels>(input["channels"]!!)
        // The fixture's trace is the stored [x, y, v] tuple shape.
        val trace = json.decodeFromJsonElement<List<List<Double>>>(input["trace"]!!)
            .map { TraceSample(x = it[0], y = it[1], v = it[2]) }

        assertEquals(
            json.decodeFromJsonElement<List<List<Limits.LimitRun>>>(expected["runs"]!!),
            channels.laps.map { Limits.limitRuns(it) },
        )
        assertEquals(
            json.decodeFromJsonElement<Limits.SessionLimits>(expected["session"]!!),
            Limits.sessionLimits(channels),
        )
        assertEquals(expected["summary"]!!.jsonPrimitive.content, Limits.limitSummary(channels))
        assertEquals(
            expected["quietSummary"]!!.jsonPrimitive.content,
            Limits.limitSummary(
                SessionChannels(
                    v = 1,
                    dStepM = 20.0,
                    laps = listOf(LapChannels(n = 1, timeMs = 0, flags = List(20) { 0.0 })),
                ),
            ),
        )
        assertEquals(JsonNull, expected["noData"])
        assertNull(
            Limits.limitSummary(SessionChannels(v = 1, dStepM = 20.0, laps = listOf(channels.laps[2]))),
        )
        assertEquals(
            json.decodeFromJsonElement<List<List<String>>>(expected["labelsAt"]!!),
            listOf(0, 3, 10).map { Limits.activeLimitLabels(channels.laps[0], it) },
        )
        assertEquals(
            json.decodeFromJsonElement<List<Limits.Marker>>(expected["markers"]!!),
            Limits.limitMarkers(channels.laps[0], channels.dStepM, trace),
        )
        assertEquals(
            json.decodeFromJsonElement<List<Limits.Marker>>(expected["noTrace"]!!),
            Limits.limitMarkers(channels.laps[0], channels.dStepM, null),
        )
    }
}
